instrumentation/kafka-clients/src/main/java/brave/kafka/TracingConsumer.java

```diff
@@ -0,0 +1,220 @@
+package brave.kafka;
+
+import brave.Span;
+import brave.Tracing;
+import brave.propagation.Propagation;
+import brave.propagation.TraceContext;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.regex.Pattern;
+import org.apache.kafka.clients.consumer.Consumer;
+import org.apache.kafka.clients.consumer.ConsumerRebalanceListener;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.clients.consumer.ConsumerRecords;
+import org.apache.kafka.clients.consumer.OffsetAndMetadata;
+import org.apache.kafka.clients.consumer.OffsetAndTimestamp;
+import org.apache.kafka.clients.consumer.OffsetCommitCallback;
+import org.apache.kafka.common.Metric;
+import org.apache.kafka.common.MetricName;
+import org.apache.kafka.common.PartitionInfo;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.header.Header;
+import zipkin.internal.Util;
+
+/**
+ * Kafka Consumer decorator. Read records headers and finish producers spans if possible.
+ */
+class TracingConsumer<K, V> implements Consumer<K, V> {
+
+  private final Tracing tracing;
+  private final Consumer<K, V> wrappedConsumer;
+
+  TracingConsumer(Tracing tracing, Consumer<K, V> consumer) {
+    this.wrappedConsumer = consumer;
+    this.tracing = tracing;
+  }
+
+  /**
+   * For each poll, finish all spans in the record list.
+   */
+  @Override
+  public ConsumerRecords<K, V> poll(long timeout) {
+    ConsumerRecords<K, V> records = wrappedConsumer.poll(timeout);
+    for (ConsumerRecord<K, V> record : records) {
+      maybeFinishProducerSpan(record);
+    }
+    return records;
+  }
+
+  /**
+   * Close the producer async span extracted from the headers.
+   */
+  private void maybeFinishProducerSpan(ConsumerRecord record) {
+    TraceContext context = tracing.propagation()
+        .extractor(getHeaders())
+        .extract(record).context();
+
+    if (context != null) {
+      Span span = tracing.tracer().joinSpan(context);
+      span.kind(Span.Kind.CONSUMER).start().finish();
+    }
```

Comment:
if you don't inject a child span onto the record, then when someone tries to resume that trace+record, they might add data to the consumer span (which already ended), right?